### PR TITLE
tools: add zsync

### PR DESCRIPTION
**Pull request by aparcar**
create tar.xz files for ImageBuilder and SDK. This saves bandwidth when daily updating local snapshots.

### CODE REVIEW

**Comment by diizzyy** [comment]
Wouldn't it make more sense to have bsdiff and/or https://github.com/thoughtpolice/minibsdiff ?
I know Debian looked into this at some point and found it to be much more efficient than zsync? I'll have a look if I can find that post again later

**Comment by aparcar** [comment]
I think that's a different approach. Using (mini)bsdiff would require to keep patches of all previous snapshot builds and also keep versions of them, so downloading clients would know *where to start*. However with zsync it compares the local file with an index (.zsync) provided by the server.

**Comment by diizzyy** [comment]
Perhaps do a weekly schedule or something, I'm not sold on using something like zsync. Something like librevault (p2p) would make more sense but that project seems dead...

**Comment by dangowrt** [comment]
@lynxis http://lists.infradead.org/pipermail/openwrt-adm/2019-February/000995.html
Should we just give it a shot?

**Comment by aparcar** [comment]
I gave it a try and it resulted in bandwidth saving of ~25%. This is somewhat poor but considering that zsync doesn't work with `tar.xz`...

The resulting .zsync file took some ~165KB for the ar71xx/generic ImageBuilder, I think that size is feasible to give it a try.

In some future someone (:tm:) could add `tar.xz` support.

**Comment by diizzyy** [comment]
I'd also be hesitant use something that's more or less dead upstream.
portsnap might be worth looking into but it's a bit FreeBSD-centric and forking it would essentially mean "no support".

https://svnweb.freebsd.org/base/head/usr.sbin/portsnap/portsnap/portsnap.sh?revision=326276&view=markup
https://en.wikipedia.org/wiki/Portsnap

First step would be to evaluate how well bspatch (bsdiff) handles xz compressed archives.

File: tools/zsync/Makefile [hunk 1]
```diff
@@ -0,0 +1,26 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=zsync
+PKG_VERSION:=f0ea808bfd957d7589420585b3d25081e0040e6d
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/ahmedammar/zsync/tar.gz/$(PKG_VERSION)?
```
Comment (lynxis) [inline tools/zsync/Makefile]: I would prefer using a git+https:// download with a hash over an external service.

Comment (aparcar) [reply tools/zsync/Makefile]: Sorry I don't understand what you mean, could you please point me to an example where it is done this way?

Comment (adrianschmutzler) [reply tools/zsync/Makefile]: I think it means: "You should not use codeload.github.com" but some other source.

**Comment by adrianschmutzler** [comment]
> I'd also be hesitant use something that's more or less dead upstream.

@aparcar Can you respond on this?

File: tools/zsync/Makefile (hunk 1 above)
Comment (neheb) [reply tools/zsync/Makefile]: local tarball basically.

There's no good solution. codeload changes the hashes from time to time (I've only seen it when using submodules) and local tarballs also change hashes in surprising ways. Many of the tarballs currently on the OpenWrt mirrors have wrong hashes. It causes no issues as it falls back to cloning the git repo (not a shallow clone oddly enough).

The latter is smaller as it uses .xz as an archive, usually.

**Comment by aparcar** [comment]
Sure let's close this, we got a cdn now anyway and the the space saving is not that great...

**Pull request closed without merging**